File: src/main/java/toast/pine/Classes/Quests/Objectives/MonsterSlaying.java
package toast.pine.Classes.Quests.Objectives;

import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDeathEvent;
import toast.pine.Classes.Quests.QuestObjective;
import toast.pine.SocialSystem.PlayerSocial;

public class MonsterSlaying implements QuestObjective, Listener {
    private final int targetQuantity;
    private int currentProgress;
    private final PlayerSocial owner;

    public MonsterSlaying(int targetQuantity, PlayerSocial playerSocial) {
        this.targetQuantity = targetQuantity;
        this.currentProgress = 0;
        this.owner = playerSocial;

    }


    @Override
    public boolean isComplete() {
        if (currentProgress >= targetQuantity) {
            EntityDeathEvent.getHandlerList().unregister(this);
            return true;
        }
        return false;
    }

    @Override
    public String getDescription() {
        return "Kill " + targetQuantity + " monsters.";
    }

    @Override
    public void updateProgress() {
        currentProgress++;
    }

    @EventHandler
    private void onPlayerKill(EntityDeathEvent event) {
        if(isComplete()) return;
        if (event.getEntity().getKiller() == owner.getPlayer()) {
            updateProgress();
        }
    }

}

File: src/main/java/pine/toast/Classes/Quests/Objectives/MonsterSlaying.kt
package toast.pine.Classes.Quests.Objectives

import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDeathEvent
import toast.pine.Classes.Quests.QuestObjective
import toast.pine.socialSystem.PlayerSocial

class MonsterSlaying(private val targetQuantity: Int, playerSocial: PlayerSocial) : QuestObjective, Listener {
    private var currentProgress = 0
    private val owner: PlayerSocial

    init {
        owner = playerSocial
    }

    override val isComplete: Boolean
        get() {
            if (currentProgress >= targetQuantity) {
                EntityDeathEvent.getHandlerList().unregister(this)
                return true
            }
            return false
        }


    override val description: String get() = "Kill $targetQuantity monsters."

    override fun updateProgress() {
        currentProgress++
    }

    @EventHandler
    private fun onPlayerKill(event: EntityDeathEvent) {
        if (isComplete) return
        if (event.entity.killer === owner.getPlayer()) {
            updateProgress()
        }
    }
}
